pallets/xcmp-queue/src/lib.rs

```diff
@@ -941,84 +510,173 @@ impl<T: Config> Pallet<T> {
 					s[index].state = OutboundState::Ok;
 				}
 			} else {
-				debug_assert!(false, "WARNING: Attempt to resume channel that was not suspended.");
+				defensive!("WARNING: Attempt to resume channel that was not suspended.");
 			}
 		});
 	}
+
+	fn enqueue_xcmp_messages(
+		sender: ParaId,
+		mut xcms: Vec<BoundedVec<u8, MaxXcmpMessageLenOf<T>>>,
+		meter: &mut WeightMeter,
+	) {
+		if meter
```

Comment:
I've been looking through the code couple of times and those drops is what worries me most. In previous implementation, we've only been using the `meter` to process messages. And now it also accounts pushing to the queue for later processing. While this is technically correct, it (it theory) can lead to accidental message drops.

For example, in `enqueue_xcmp_messages` benchmark, we are using maximal possible message size. So the weight returned is the maximal possible. But what if instead of largest possible messages, the sending chain sends a bunch of smallest possible messages? Then, it the size effect is too large, then we can drop it here, because `T::WeightInfo::enqueue_xcmp_messages(<nuymbe-of-small-messages>)` is large enough to fit into the `meter`.

I know that it is unlikely to happen, because we have `1/4` of block to enqueue messages. But if there are many inbound HRMP channels (I could imaging bridge hub will have many) and a lot of inbound messages, it can happen.

I also think that we probably can ignore the overweight problem here and simply return actual used weight, because IIUC it is called from the inherent (`set_validation_data`).

That said, I'm not a pro here - it just caught my eye :)

Reply:
> We now use the weight only for enqueueing and signal processing - not message processing anymore.

True and that's what is worrying me :) Like - previously we have been ignoring this weight and now we don't (which is technically correct). And because of that we may drop some messages.

But ok - let's go with he current impl. I think there's some significant reserve (block quarter) for just enqueueing.